src/functional/otf/compilation/build_data.py

```diff
@@ -0,0 +1,79 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import dataclasses
+import enum
+import json
+import pathlib
+from typing import Final, Optional
+
+
+_DATAFILE_NAME: Final = "gt4py.json"
+
+BuildStatus = enum.IntEnum("BuildStatus", ["STARTED", "CONFIGURED", "COMPILED"], start=0)
+
+
+@dataclasses.dataclass(frozen=True)
+class BuildData:
+    """
+    Tracks build status and results of GT4Py program compilations.
+
+    Build systems projects should keep this information up to date for each build,
+    stored to a json file in the build folder at all times.
+    """
+
+    status: BuildStatus
+    module: pathlib.Path
+    entry_point_name: str
+
+    def to_json(self) -> dict[str, str]:
+        return dataclasses.asdict(self) | {"status": self.status.name, "module": str(self.module)}
+
+    @classmethod
+    def from_json(cls, data) -> BuildData:
+        return cls(
+            **(
+                data
+                | {
+                    "status": getattr(BuildStatus, data["status"]),
+                    "module": pathlib.Path(data["module"]),
+                }
+            )
+        )
+
+
+def contains_data(path: pathlib.Path) -> bool:
+    return (path / _DATAFILE_NAME).exists()
+
+
+def read_data(path) -> Optional[BuildData]:
+    if contains_data(path):
+        return BuildData.from_json(json.loads((path / _DATAFILE_NAME).read_text()))
```

Comment:
In which case an error will be raised, which is perfectly fine, given that something really, really weird just happened.

Reply:
But it's still incorrect in that corner case, because it should return `None` instead of throwing an exception. It would be better like
```python
    try:
      BuildData.from_json(json.loads((path / _DATAFILE_NAME).read_text()))
    except FileNotFoundError:
      return None
```

Reply:
It is not incorrect because a deleted build data file means there was an error in a build rather than no build has been attempted. 

In both cases below there is a point in time where the build data file does not exist, but they are not the same case.
- No build attempted -> should return None
- build started and build data got lost -> should raise Error

That said this corner case should be rare and was not a design consideration so I have nothing about the try-except version which neglects it.